core/services/keeper/orm.go

```diff
@@ -111,18 +133,34 @@ WHERE
 			upkeep_registrations.last_run_block_height + $2 < $3 AND
 			upkeep_registrations.last_run_block_height < ($3 - ($3 % keeper_registries.block_count_per_turn))
 		)
-	) AND
-	keeper_registries.keeper_index = (
-		upkeep_registrations.positioning_constant + (($3 - ($3 % keeper_registries.block_count_per_turn)) / keeper_registries.block_count_per_turn)
-	) % keeper_registries.num_keepers
-ORDER BY upkeep_registrations.id ASC, upkeep_registrations.upkeep_id ASC
+	)
+	AND 
+   -- this keeper's turn AND last perform not by this keeper
+    (
+                keeper_registries.keeper_index = ((CAST(upkeep_registrations.positioning_constant AS bit(32)) #
+                                                   CAST($4 AS bit(32)))::int % keeper_registries.num_keepers)
+            AND
+                (upkeep_registrations.last_keeper_index != keeper_registries.keeper_index OR last_keeper_index IS NULL)
+        )
+   OR
+   -- next keeper's turn AND last perform was by next keeper so cover
+    (
+                    (keeper_registries.keeper_index + 1) % keeper_registries.num_keepers =
+                    ((CAST(upkeep_registrations.positioning_constant AS bit(32)) #
+                      CAST($4 AS bit(32)))::int % keeper_registries.num_keepers)
```

Comment:
We can probably use macros / some other sql construct to define some formulas, remove duplication here and make the whole query readable. Would be great if you can do some research into it